File: NLPL/Parser.cpp
#include "stdafx.h"
#include "Parser.h"


using namespace nldb;

Parser::Parser()
{
	this->split_expr = "\\s+";
	this->remove_expr = "[^à-ÿÀ-ß¸¨a-zA-Z0-9\\s\\t\\r\\n\\!\\?\\,\\;\\:\\%\\$\\#\\@\\*\\+\\-\\&\\^\\(\\)\\[\\]\\{\\}\\=\\<\\>]";

}

Parser::~Parser()
{
}

Parser::Parser(const string& expr)
	: split_expr(expr)
{ }

vector<string> Parser::Split(const string& text)
{
	vector<string> substr;
	regex rgx(split_expr);
	sregex_token_iterator iter(text.begin(), text.end(), rgx, -1);
	sregex_token_iterator end;
	for (; iter != end; ++iter)
		substr.push_back(*iter);
	return substr;
}

string Parser::Normilize(const string& text)
{
	string result;
	transform(text.begin(), text.end(), result.begin(), ::tolower);
	return result;
}

File: NLDB.Core/Parser.cs
using System;
using System.Text.RegularExpressions;

namespace NLDB
{
    //TODO: Создать модульные тесты для Parser
    /// <summary>
    /// Класс, используемый для иерархической токенизации текста.
    /// Иерархическая токенизация - такое разбиение текста на врагменты, в котором каждый фрагмент имеет свой идентификатор и,
    /// либо состоит из других фрагментов, либо является неделимым фрагментом текста (символом)
    /// </summary>
    [Serializable]
    public class Parser
    {
        //Служебные константы, используемые для подстановки в текст для сокращения общего количества токенов по тексту
        private const string specSymbolNumber = "{ХХХ}";
        private const string specSymbolRomeNumber = "{ХХХ}";
        private const string specSymbolEnglish = "{англ}";
        private const string specSymbolShorthand = "";

        private readonly string SplitExpr;
        //private string RemoveExpr = "";
        private readonly Regex splitRegex;
        private static Regex removeRegex = new Regex(@"[^а-яА-ЯёЁ\d\s\n\!\.\,\;\:\-\{\}\=\<\>\""\p{Pd}]", RegexOptions.Compiled);
        private static Regex replaceNumbersRegex = new Regex(@"\b\d+((\.|\,)\d+)?", RegexOptions.Compiled);
        private static Regex replaceEnglishRegex = new Regex(@"[a-zA-Z]+", RegexOptions.Compiled);
        private static Regex removeShorthands = new Regex(@"\b[а-яА-ЯёЁ]\s?\.", RegexOptions.Compiled);

        public Parser(string splitExpr)
        {
            SplitExpr = splitExpr;
            splitRegex = new Regex(SplitExpr, RegexOptions.Compiled);
        }

        public Parser()
        {
            SplitExpr = "";
            splitRegex = new Regex(SplitExpr, RegexOptions.Compiled);
        }

        public string[] Split(string text)
        {
            return splitRegex.Split(text);
        }

        public static string Normilize(string text)
        {
            text = removeRegex.Replace(text.ToLower().Trim(), "");
            text = removeShorthands.Replace(text, specSymbolShorthand);
            text = replaceNumbersRegex.Replace(text, specSymbolNumber);
            text = replaceEnglishRegex.Replace(text, specSymbolEnglish);
            return text;
        }

    }
}
